src/lib/crypto/kimchi_pasta_snarky_backend/plonk_constraint_system.ml

```diff
@@ -2420,4 +2441,36 @@ end = struct
               Option.try_with (fun () -> reduce_to_v values.(i)) )
         in
         add_row sys values kind coeffs
+
+  (* ((Fp.t * V.t) list * Fp.t option) *)
+  type concrete_table = ((Fp.t * V.t) list * Fp.t option) Internal_var.Table.t
+  [@@deriving bin_io]
+
+  (* V.t option array list *)
+  type concrete_rows_rev = V.t option array list [@@deriving bin_io]
+
+  let dump_extra_circuit_data (sys : t) base_path =
+    let rows_rev_name = base_path ^ "_rows_rev.bin" in
+    let internal_vars_name = base_path ^ "_internal_vars.bin" in
+    let gates_json_name = base_path ^ "_gates.json" in
+    if Sys.file_exists rows_rev_name then Sys.remove rows_rev_name ;
+    if Sys.file_exists internal_vars_name then Sys.remove internal_vars_name ;
+    if Sys.file_exists gates_json_name then Sys.remove gates_json_name ;
+
+    let table : concrete_rows_rev = sys.rows_rev in
+    let size = bin_size_concrete_rows_rev table in
+    let buf = Bigstring.create size in
+    ignore (bin_write_concrete_rows_rev buf ~pos:0 table : int) ;
+    Core_kernel.Out_channel.write_all rows_rev_name
+      ~data:(Bigstring.to_string buf) ;
+
+    let table : concrete_table = sys.internal_vars in
+    let size = bin_size_concrete_table table in
+    let buf = Bigstring.create size in
+    ignore (bin_write_concrete_table buf ~pos:0 table : int) ;
+    Core_kernel.Out_channel.write_all internal_vars_name
```

Comment:
I am assuming there's already a parser in place to recover the original constraint system from these jsons?